include/effectmanager.h

```diff
@@ -96,6 +95,61 @@ class EffectManager : public IJSONSerializable
         }
     }
 
+    void LoadJSONAndMissingEffects(const JsonArrayConst& effectsArray)
+    {
+        std::set<int> loadedEffectNumbers;
+
+        // Create effects from JSON objects, using the respective factories in g_EffectFactories
+        auto& jsonFactories = g_EffectFactories.GetJSONFactories();
+
+        for (auto effectObject : effectsArray)
+        {
+            int effectNumber = effectObject[PTY_EFFECTNR];
+            auto factoryEntry = jsonFactories.find(effectNumber);
+
+            if (factoryEntry == jsonFactories.end())
+                continue;
+
+            auto pEffect = factoryEntry->second(effectObject);
+
+            if (pEffect != nullptr)
+            {
+                _vEffects.push_back(pEffect);
+                loadedEffectNumbers.insert(effectNumber);
+            }
+        }
+
+        // Now add missing effects from the default factory list
+        auto &defaultFactories = g_EffectFactories.GetDefaultFactories();
+
+        // We iterate manually, so we can use where we are as the starting point for a later inner loop
+        for (auto iter = defaultFactories.begin(); iter != defaultFactories.end(); iter++)
+        {
+            int effectNumber = iter->EffectNumber;
+
+            // If we've already loaded this effect (number) from JSON, we can move on to check the next one
+            if (loadedEffectNumbers.count(effectNumber))
+                continue;
+
+            // We found an effect (number) in the default list that we have not yet loaded from JSON.
+            //   So, we go through the rest of the default factory list to create and add to our effects
+            //   list all instances of this effect.
+            std::for_each(iter, defaultFactories.end(), [&](const EffectFactories::NumberedFactory& numberedFactory)
+                {
+                    if (numberedFactory.EffectNumber != effectNumber)
+                        return;
+
+                    auto pEffect = numberedFactory.Factory();

```

Comment:
auto &

Reply:
No, that also won't compile. In this case I believe it is because `numberedFactory.Factory()` returns a `shared_ptr` that is created in the function.